src/Microsoft.VisualStudio.Editor.Razor/DefaultVisualStudioRazorParser.cs

```diff
@@ -0,0 +1,372 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.Language.Legacy;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Language.Intellisense;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Operations;
+using ITextBuffer = Microsoft.VisualStudio.Text.ITextBuffer;
+using Timer = System.Threading.Timer;
+
+namespace Microsoft.VisualStudio.Editor.Razor
+{
+    internal class DefaultVisualStudioRazorParser : VisualStudioRazorParser, IDisposable
+    {
+        // Internal for testing.
+        internal TimeSpan IdleDelay = TimeSpan.FromSeconds(3);
+        internal Timer _idleTimer;
+
+        private readonly object IdleLock = new object();
+        private readonly ICompletionBroker _completionBroker;
+        private readonly IEnumerable<ICanHasContextChangedListener> _parserContextChangedListeners;
+        private readonly VisualStudioDocumentTracker _documentTracker;
+        private readonly ForegroundDispatcher _dispatcher;
+        private readonly RazorTemplateEngineFactoryService _templateEngineFactory;
+        private readonly string _projectDirectory;
+        private readonly ErrorReporter _errorReporter;
+        private BackgroundParser _parser;
+        private RazorSyntaxTreePartialParser _partialParser;
+        private BraceSmartIndenter _braceSmartIndenter;
+        private RazorTemplateEngine _templateEngine;
+        private RazorCodeDocument _codeDocument;
+        private ITextSnapshot _snapshot;
+
+        // For testing only
+        internal DefaultVisualStudioRazorParser(RazorCodeDocument codeDocument)
+        {
+            _codeDocument = codeDocument;
+        }
+
+        public DefaultVisualStudioRazorParser(
+            ForegroundDispatcher dispatcher,
+            VisualStudioDocumentTracker documentTracker,
+            RazorTemplateEngineFactoryService templateEngineFactory,
+            string filePath,
+            string projectDirectory,
+            ErrorReporter errorReporter,
+            ICompletionBroker completionBroker,
+            IEditorOperationsFactoryService editorOperationsFactory,
+            IEnumerable<ICanHasContextChangedListener> parserContextChangedListeners)
+        {
+            if (dispatcher == null)
+            {
+                throw new ArgumentNullException(nameof(dispatcher));
+            }
+
+            if (documentTracker == null)
+            {
+                throw new ArgumentNullException(nameof(documentTracker));
+            }
+
+            if (templateEngineFactory == null)
+            {
+                throw new ArgumentNullException(nameof(templateEngineFactory));
+            }
+
+            if (string.IsNullOrEmpty(filePath))
+            {
+                throw new ArgumentException(Resources.ArgumentCannotBeNullOrEmpty, nameof(filePath));
+            }
```

Comment:
Should this be part of the tracker?

Reply:
It could be I'd need to look into how that information gets refreshed when the document moves and when that information is available (today we get it from the Roslyn text document associated with the buffer).

Reply:
Investigated this and can definitely track that data on the document tracker. Also found that we don't track document movements correctly in our document tracker: https://github.com/aspnet/Razor/issues/1725